### PR TITLE
PS2 compatibility changes - WIP

### PR DESCRIPTION
**Pull request by darrena092**
This fixes a few issues I've found:

- Acknowledging fast SPI clocks on PS2 meant for other peripherals, due to the speed of the pico. The PIO programs run slower now.
- Random ACKs. This was due to how the ACK signal was triggered. It now won't trigger unless it counts 8 clocks.
- Transaction restarts. Fixed this by re-structuring the memory card simulator to work as a state machine. 

I do have one problem that I've yet to solve though, which is why this is still in draft. Occasionally both the PS1 and PS2 seem to think that the card has been disconnected.

### CODE REVIEW

**Comment by darrena092** [comment]
Relates to: #2

File: src/memcard_simulator_sm.c [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+                    next_state = MC_END;
+                }
+            }
+            sm_byte_counter++;
+        break;
+        case MC_EXECUTE_WRITE: // do a write operation
+            if(memory_card_is_sector_valid(&mc, sm_address)) {
+                uint8_t* sec_ptr = memory_card_get_sector_ptr(&mc, sm_address);
+                if(sm_byte_counter == 0) {
+                    checksum = ((sm_address & 0xFF00) << 8) ^ (sm_address & 0x00FF);
```
Comment (dangiu) [inline src/memcard_simulator_sm.c:250]:
```suggestion
                    checksum = ((sm_address & 0xFF00) >> 8) ^ (sm_address & 0x00FF);
```
Again shift to the right is required to isolate MSB

File: src/memcard_simulator_sm.c [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    pio_sm_clear_fifos(pio, smDatWriter);
+    pio_interrupt_clear(pio0, 0);
+    pio_enable_sm_mask_in_sync(pio, 1 << smCmdReader | 1 << smDatWriter);
+
+    // Reset mc state machine
+    current_state = MC_IDLE;
+    next_state = MC_IDLE;
+    command_state = MC_IDLE;
+    sm_byte_counter = 0;
+    sm_address = 0x00;
```
Comment (dangiu) [inline src/memcard_simulator_sm.c:77]:
```suggestion
    sm_address = 0x0000;
```
For consistency to highlight the size

File: src/memcard_simulator_sm.c [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    bool valid_command = false;
+    current_state = next_state;
+
+    switch(current_state) {
+        case MC_IDLE: // idle / sleeping
+            sm_byte_counter = 0;
+            command_state = MC_IDLE;
+            checksum = 0x00;
+            recv_checksum = 0x00;
+            sm_address = 0x00;
```
Comment (dangiu) [inline src/memcard_simulator_sm.c:130]:
```suggestion
            sm_address = 0x0000;
```

**Review by dangiu — CHANGES_REQUESTED** [review]
I like how you restructured the code.
At times the byte_counter thingy seems a bit confusing (mostly in the READ/WRITE states where you have to subtract stuff to get the right index) but right know I'm not seeing an obvious way to improve it.
The END state is also a bit confusing, mostly for the fact that it compares the received checksum against the computed one which only makes sense on write commands. I've noticed that in read commands you reset the checksum before transitioning to END state which gets the desired behavior but is a bit tricky to understand.

The changes highlighted might be enough to get rid of the spontaneous errors. I don't think that fixing the ID command will improve on anything... Until now I've never been able to capture an instance where the PSX uses that command but I only have another Pico as oscilloscope so my sniffing capabilities are really limited.
I think that some of the problems could have been related to high sectors read/write, sectors above 255 could not be reached due to the incorrect type of sm_address.

I've not been able to test these fixes... Let me know if they make sense at all or if I'm missing something :)

File: src/memcard_simulator_sm.c [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+                    write_byte_blocking(pio, smDatWriter, data);
+                }
+
+                next_state = command_state;
+                command_state = MC_IDLE;
+                sm_byte_counter = 0;
+            }
+        break;
+        case MC_EXECUTE_ID: // send mc id - used to identify which type of device this is
+            if(sm_byte_counter++ < sizeof(id_data)) {
```
[the remark this answers is not in the document]
Comment (darrena092) [reply src/memcard_simulator_sm.c]: Done

File: src/memcard_simulator_sm.c [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+                write_byte_blocking(pio, smDatWriter, id_data[sm_byte_counter]);
```
[the remark this answers is not in the document]
Comment (darrena092) [reply src/memcard_simulator_sm.c]: Done

File: src/memcard_simulator_sm.c [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    MC_END
+};
+
+uint8_t current_state = MC_IDLE;
+uint8_t next_state = MC_IDLE;
+uint8_t command_state = MC_IDLE;
+uint8_t checksum = 0x00;
+uint8_t recv_checksum = 0x00;
+uint8_t sm_byte_counter = 0;
+uint8_t sm_address = 0x00;
```
[the remark this answers is not in the document]
Comment (darrena092) [reply src/memcard_simulator_sm.c]: Done

File: src/memcard_simulator_sm.c [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+                memory_card_update_timestamp(&mc);
+                next_state = MC_IDLE;
+                in_transaction = false;
+            }
+        break;
+        case MC_EXECUTE_READ: // do a read operation
+            if(sm_byte_counter == 0) {
+                // Send ACK2
+                write_byte_blocking(pio, smDatWriter, MC_ACK2);
+                checksum = ((sm_address & 0xFF00) << 8) ^ (sm_address & 0x00FF);
```
[the remark this answers is not in the document]
Comment (darrena092) [reply src/memcard_simulator_sm.c]: Done

File: src/memcard_simulator_sm.c (hunk 1 above)
Comment (darrena092) [reply src/memcard_simulator_sm.c]: Done

File: src/memcard_simulator_sm.c (hunk 2 above)
Comment (darrena092) [reply src/memcard_simulator_sm.c]: Done

File: src/memcard_simulator_sm.c (hunk 3 above)
Comment (darrena092) [reply src/memcard_simulator_sm.c]: Done

**Comment by darrena092** [comment]
That's the things you mentioned fixed, about to do more testing. On the ID command, interestingly enough the PS2 does use it. And not implementing it makes the PS2 think that a pocketstation is connected!

**Comment by darrena092** [comment]
Still not ready to go yet. Fixing those issues you highlighted solved the weirdness around writes, but there's still an issue with syncing. With that `in_transaction` flag, it seems to just never sync, with removing it, it'll go into a sync right in the middle of transaction. More work needed, but I think that's me done for today!

**Comment by darrena092** [comment]
Closing this, going to take a different approach

**Pull request closed without merging**